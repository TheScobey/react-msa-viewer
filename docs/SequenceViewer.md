# `SequenceViewer` (component)



## Props

### `showModBar`

defaultValue: `true`

