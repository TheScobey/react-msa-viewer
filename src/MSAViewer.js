import React, { Component } from 'react';
import PropTypes from 'prop-types';

const schemes = new (require('msa-colorschemes'))();

class MSAViewer extends Component {

  constructor(props) {
    super(props);
    this.canvas = React.createRef();
  }

  componentDidMount() {
    this.ctx = this.canvas.current.getContext('2d');
    this.scheme = schemes.getScheme(this.props.scheme);
    this.drawSequences();
  }

  drawSequences() {
    const xSize = 20;
    const ySize = 20;
    let yPos = 0, xPos = 0;
    for (const seq of this.props.sequences) {
      for (const el of seq.sequence) {
        this.ctx.font = "20px Arial";
        this.ctx.fillStyle = this.scheme.getColor(el);
        this.ctx.fillRect(xPos, yPos, xSize, ySize);
        this.ctx.fillStyle = "#000000";
        this.ctx.fillText(el, xPos + 2, yPos + 17);
        xPos += xSize;
      }
      xPos = 0;
      yPos += ySize;
    }
  }

  componentDidUpdate() {

  }

  shouldComponentUpdate(newProps) {
    return this.props.target !== newProps.target;
  }

  render() {
    return (
      <canvas ref={this.canvas} with={500} height={500}>
      </canvas>
    );
  }
}

const SequencePropType = PropTypes.shape({
  name: PropTypes.string,
  sequence: PropTypes.string,
})

MSAViewer.propTypes = {
  color: PropTypes.string,
  scheme: PropTypes.string,
  sequences: PropTypes.arrayOf(SequencePropType).isRequired,
};

MSAViewer.defaultProps = {
  color: "red",
  scheme: "clustal",
}

export {MSAViewer};
export default MSAViewer;
