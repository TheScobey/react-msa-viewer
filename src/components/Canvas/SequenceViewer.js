/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import PropTypes from 'prop-types';

import {
  floor,
  clamp,
} from 'lodash-es';

import DraggingComponent from './DraggingComponent';
import TilingGrid from './CanvasTilingGrid';
import CanvasCache from './CanvasCache';

import { updatePosition } from '../../store/positionReducers';
import positionStoreMixin from '../../store/positionStoreMixin';
import msaConnect from '../../store/connect'

import Mouse from '../../utils/mouse';
import { roundMod } from '../../utils/math';

// TODO: maybe move into the store
class SequenceViewerComponent extends DraggingComponent {

  constructor(props) {
    super(props);
    this.tileCache = new CanvasCache();
    this.residueTileCache = new CanvasCache();
    this.tilingGridManager = new TilingGrid();
  }

  drawScene() {
    const positions = this.getTilePositions();
    const elements = this.drawTiles(positions);
  }

  getTilePositions() {
    const startXTile = Math.max(0, this.position.currentViewSequencePosition - this.props.cacheElements);
    const startYTile = Math.max(0, this.position.currentViewSequence - this.props.cacheElements);
    const endYTile = Math.min(this.props.sequences.length,
      startYTile + this.props.nrYTiles + 2 * this.props.cacheElements,
    );
    const endXTile = Math.min(this.props.sequences.maxLength,
      startXTile + this.props.nrXTiles + 2 * this.props.cacheElements,
    );
    return {startXTile, startYTile, endXTile, endYTile};
  }

  renderTile = ({row, column}) => {
    const key = row + "-" + column;
    return this.tileCache.createTile({
      key: key,
      tileWidth: this.props.tileWidth * this.props.xGridSize,
      tileHeight: this.props.tileHeight * this.props.yGridSize,
      create: ({canvas}) => {
        this.tilingGridManager.draw({
          ctx: canvas,
          sequences:this.props.sequences,
          colorScheme:this.props.colorScheme,
          tileFont:this.props.tileFont,
          tileHeight:this.props.tileHeight,
          tileWidth:this.props.tileWidth,
          startYTile:row,
          startXTile:column,
          residueTileCache:this.residueTileCache,
          endYTile:row + this.props.yGridSize,
          endXTile:column + this.props.xGridSize,
        });
      },
    });
  }


  drawTiles({startXTile, startYTile, endXTile, endYTile}) {
    const elements = [];
    const xGridSize = this.props.xGridSize;
    const yGridSize = this.props.yGridSize;
    startYTile = roundMod(startYTile, yGridSize);
    startXTile = roundMod(startXTile, xGridSize);

    const yOffset = (this.position.currentViewSequence - startYTile) * this.props.tileHeight - this.position.yPosOffset
    const xOffset = (this.position.currentViewSequencePosition - startXTile) * this.props.tileWidth - this.position.xPosOffset;

    // TODO: cut-off end tiles
    for (let i = startYTile; i < endYTile; i = i + yGridSize) {
      for (let j = startXTile; j < endXTile; j = j + xGridSize) {
        const canvas = this.renderTile({row: i, column: j, canvas: this.ctx});
        const width = xGridSize * this.props.tileWidth;
        const height = yGridSize * this.props.tileHeight;
        const yPos = i * this.props.tileHeight - yOffset;
        const xPos = j * this.props.tileWidth - xOffset;
        this.ctx.ctx.drawImage(canvas, xPos, yPos, width, height);
      }
    }
  }

  onPositionUpdate = (oldPos, newPos) => {
    const relativeMovement = {
      xMovement: oldPos[0] - newPos[0],
      yMovement: oldPos[1] - newPos[1],
    };
    this.context.positionMSAStore.dispatch(updatePosition(relativeMovement));
  }

  positionToSequence(pos) {
    const sequences = this.props.sequences.raw;
    const seqNr = clamp(floor((this.position.yPos + pos.yPos) / this.props.tileHeight), 0, sequences.length - 1);
    const sequence = sequences[seqNr];

    const position = clamp(floor((this.position.xPos + pos.xPos) / this.props.tileWidth), 0, sequence.sequence.length - 1);
    return {
      i: seqNr,
      sequence,
      position,
      residue: sequence.sequence[position],
    }
  }

  updateScrollPosition = () => {
    this.draw();
  }

  componentDidUpdate() {
    // TODO: smarter updates
    this.draw();
  }

  /**
   * Returns the position of the mouse position relative to the sequences
   */
  currentPointerPosition(e) {
    const [x, y] = Mouse.rel(e);
    return this.positionToSequence({
      xPos: x,
      yPos: y,
    });
  }

  /**
   * Only sends an event if the actual function is set.
   */
  sendEvent(name, data) {
    if (this.props[name] !== undefined) {
      this.props[name](data);
    }
  }

  onMouseMove = (e) => {
    if (typeof this.dragFrame === "undefined") {
      //if (this.props.onResidueMouseEnter !== undefined ||
          //this.props.onResidueMouseLeave !== undefined) {
        //const eventData = this.currentPointerPosition(e);
        //const lastValue = this.currentMouseSequencePosition;
        //if (!isEqual(lastValue, eventData)) {
          //if (lastValue !== undefined) {
            //this.sendEvent('onResidueMouseLeave', lastValue);
          //}
          //this.currentMouseSequencePosition = eventData;
          //this.sendEvent('onResidueMouseEnter', eventData);
        //}
      //}
    }
    super.onMouseMove(e);
  }

  onMouseLeave = (e) => {
    //this.sendEvent('onResidueMouseLeave', this.currentMouseSequencePosition);
    this.currentMouseSequencePosition = undefined;
    super.onMouseLeave(e);
  }

  onClick = (e) => {
    //const eventData = this.currentPointerPosition(e);
    //this.sendEvent('onResidueClick', eventData);
    super.onClick(e);
  }

  onDoubleClick = (e) => {
    const eventData = this.currentPointerPosition(e);
    this.sendEvent('onResidueDoubleClick', eventData);
    super.onDoubleClick(e);
  }

  //shouldComponentUpdate(newProps) {
    //// TODO: check recursively
    ////return this.props.target !== newProps.target;
    //return true;
  //}

  // to make react-docgen happy
  render() {
    return super.render();
  }
}

positionStoreMixin(SequenceViewerComponent, {withX: true, withY: true});

SequenceViewerComponent.defaultProps = {
  showModBar: false,
  xGridSize: 10,
  yGridSize: 10,
  cacheElements: 10,
};

SequenceViewerComponent.propTypes = {
  /**
   * Show the custom ModBar
   */
  showModBar: PropTypes.bool,

  /**
   * Callback fired when the mouse pointer is entering a residue.
   */
  onResidueMouseEnter: PropTypes.func,

  /**
   * Callback fired when the mouse pointer is leaving a residue.
   */
  onResidueMouseLeave: PropTypes.func,

  /**
   * Callback fired when the mouse pointer clicked a residue.
   */
  onResidueClick: PropTypes.func,

  /**
   * Callback fired when the mouse pointer clicked a residue.
   */
  onResidueDoubleClick: PropTypes.func,

  xGridSize: PropTypes.number.isRequired,
  yGridSize: PropTypes.number.isRequired,
};

const mapStateToProps = state => {
  // Fallback to a smaller size if the given area is too large
  const width = Math.min(
    state.props.width,
    state.sequences.maxLength * state.props.tileWidth
  );
  const height = Math.min(
    state.props.height,
    state.sequences.length * state.props.tileHeight
  );
  return {
    sequences: state.sequences,
    width,
    height,
    tileWidth: state.props.tileWidth,
    tileHeight: state.props.tileHeight,
    tileFont: state.props.tileFont,
    msecsPerFps: state.props.msecsPerFps,
    colorScheme: state.props.colorScheme,
    engine: state.props.engine,
    nrXTiles: state.sequenceStats.nrXTiles,
    nrYTiles: state.sequenceStats.nrYTiles,
  }
}

//const mapDispatchToProps = dispatch => {
  //return {
    //updatePosition: flow(updatePosition, dispatch),
  //}
//}

export default msaConnect(
  mapStateToProps,
  //mapDispatchToProps,
)(SequenceViewerComponent);
