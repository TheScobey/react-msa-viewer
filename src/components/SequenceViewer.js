/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import msaConnect from '../store/connect'
import { updatePosition } from '../store/actions'
import PropTypes from 'prop-types';

import {
  flow,
  floor,
  clamp,
  isEqual,
} from 'lodash-es';

import DraggingComponent from './DraggingComponent';

import Mouse from '../utils/mouse';

// TODO: maybe move into the store
class SequenceViewerComponent extends DraggingComponent {

  /**
   * Draws the currently visible sequences.
   * Called on every sequence movement.
   */
  drawScene() {
    const sequences = this.props.sequences.raw;
    const tileWidth = this.props.tileWidth;
    const tileHeight = this.props.tileHeight;
    const xInitPos = -(this.props.position.xPos % tileWidth);
    let yPos = -(this.props.position.yPos % tileHeight);
    let i = this.currentViewSequence();
    for (; i < sequences.length; i++) {
      const sequence = sequences[i].sequence;
      let xPos = xInitPos;
      let j = this.currentViewSequencePosition(sequence);
      for (; j < sequence.length; j++) {
        const el = sequence[j];
        this.ctx.font(this.props.tileFont);
        this.ctx.fillStyle(this.props.colorScheme.getColor(el));
        this.ctx.globalAlpha(0.7);
        this.ctx.fillRect(xPos, yPos, tileWidth, tileHeight);
        this.ctx.fillStyle("#000000");
        this.ctx.globalAlpha(1.0);
        // TODO: center the font tile
        this.ctx.fillText(el, xPos, yPos, tileWidth, tileHeight);
        xPos += tileWidth;
        if (xPos > this.props.width)
            break;
      }
      yPos += tileHeight;
      if (yPos > this.props.height)
          break;
    }
  }

  onPositionUpdate(oldPos, newPos) {
    // TODO: move this into a redux action
    const pos = this.props.position;
    pos.xPos += oldPos[0] - newPos[0];
    pos.yPos += oldPos[1] - newPos[1];
    // TODO: need maximum of sequence lengths here
    const maximum = this.props.sequences.maxLength;
    const maxWidth = maximum * this.props.tileWidth - this.props.width;
    pos.xPos = clamp(pos.xPos, 0, maxWidth);
    const maxHeight = this.props.sequences.raw.length * this.props.tileHeight - this.props.height;
    pos.yPos = clamp(pos.yPos, 0, maxHeight);
    this.props.updatePosition(pos);
  }

  // TODO: move into the reducer
  /**
   * Returns the first visible sequence on the current viewpoint.
   * Might only be partially visible.
   */
  currentViewSequence() {
    return clamp(floor(this.props.position.yPos / this.props.tileHeight), 0, this.props.sequences.length - 1)
  }

  /**
   * Returns the first visible position on the current viewpoint.
   * Might only be partially visible.
   */
  currentViewSequencePosition(sequence) {
    return clamp(floor(this.props.position.xPos / this.props.tileWidth), 0, sequence.length - 1);
  }

  positionToSequence(pos) {
    const sequences = this.props.sequences.raw;
    const seqNr = clamp(floor((this.props.position.yPos + pos.yPos) / this.props.tileHeight), 0, sequences.length - 1);
    const sequence = sequences[seqNr];

    const position = clamp(floor((this.props.position.xPos + pos.xPos) / this.props.tileWidth), 0, sequence.sequence.length - 1);
    return {
      i: seqNr,
      sequence,
      position,
      residue: sequence.sequence[position],
    }
  }

  componentDidUpdate() {
    // TODO: smarter updates
    this.draw();
  }

  // TODO: move into the redux store
  /**
   * Only fire an event if its value has changed.
   */
  fireEvent(eventName, e) {
  }

  /**
   * Returns the position of the mouse position relative to the sequences
   */
  currentPointerPosition(e) {
    const [x, y] = Mouse.rel(e);
    return this.positionToSequence({
      xPos: x,
      yPos: y,
    });
  }

  /**
   * Only sends an event if the actual function is set.
   */
  sendEvent(name, data) {
    if (this.props[name] !== undefined) {
      this.props[name](data);
    }
  }

  onMouseMove = (e) => {
    if (typeof this.dragFrame === "undefined") {
      if (this.props.onResidueMouseEnter !== undefined ||
          this.props.onResidueMouseLeave !== undefined) {
        const eventData = this.currentPointerPosition(e);
        const lastValue = this.currentMouseSequencePosition;
        if (!isEqual(lastValue, eventData)) {
          if (lastValue !== undefined) {
            this.sendEvent('onResidueMouseLeave', lastValue);
          }
          this.currentMouseSequencePosition = eventData;
          this.sendEvent('onResidueMouseEnter', eventData);
        }
      }
    }
    super.onMouseMove(e);
  }

  onMouseLeave = (e) => {
    this.sendEvent('onResidueMouseLeave', this.currentMouseSequencePosition);
    this.currentMouseSequencePosition = undefined;
    super.onMouseLeave(e);
  }

  onClick = (e) => {
    const eventData = this.currentPointerPosition(e);
    this.sendEvent('onResidueClick', eventData);
    super.onClick(e);
  }

  onDoubleClick = (e) => {
    const eventData = this.currentPointerPosition(e);
    this.sendEvent('onResidueDoubleClick', eventData);
    super.onDoubleClick(e);
  }

  //shouldComponentUpdate(newProps) {
    //// TODO: check recursively
    ////return this.props.target !== newProps.target;
    //return true;
  //}

  // to make react-docgen happy
  render() {
    return super.render();
  }
}

SequenceViewerComponent.defaultProps = {
  showModBar: true,
};

SequenceViewerComponent.PropTypes = {
  /**
   * Show the custom ModBar
   */
  showModBar: PropTypes.boolean,

  /**
   * Callback fired when the mouse pointer is entering a residue.
   */
  onResidueMouseEnter: PropTypes.func,

  /**
   * Callback fired when the mouse pointer is leaving a residue.
   */
  onResidueMouseLeave: PropTypes.func,

  /**
   * Callback fired when the mouse pointer clicked a residue.
   */
  onResidueClick: PropTypes.func,

  /**
   * Callback fired when the mouse pointer clicked a residue.
   */
  onResidueDoubleClick: PropTypes.func,
};

const mapStateToProps = state => {
  // Fallback to a smaller size if the given area is too large
  const width = Math.min(
    state.props.width,
    state.sequences.maxLength * state.props.tileWidth
  );
  const height = Math.min(
    state.props.height,
    state.sequences.length * state.props.tileHeight
  );
  return {
    position: state.position,
    sequences: state.sequences,
    width,
    height,
    tileWidth: state.props.tileWidth,
    tileHeight: state.props.tileHeight,
    tileFont: state.props.tileFont,
    msecsPerFps: state.props.msecsPerFps,
    colorScheme: state.props.colorScheme,
    engine: state.props.engine,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    updatePosition: flow(updatePosition, dispatch),
  }
}

export default msaConnect(
  mapStateToProps,
  mapDispatchToProps,
)(SequenceViewerComponent);
