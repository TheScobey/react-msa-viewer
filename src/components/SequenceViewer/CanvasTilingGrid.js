/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import CanvasComponent from '../Canvas/CanvasComponent';

class CanvasCache {
  constructor(g) {
    this.cache = {};
    this.cacheHeight = 0;
    this.cacheWidth = 0;
  }

  // returns a cached canvas
  getFontTile(letter, width, height, font) {
    // validate cache
    if (width !== this.cacheWidth || height !== this.cacheHeight || font !== this.font) {
      this.updateDimensions(width, height);
      this.font = font;
    }

    if (this.cache[letter] === undefined) {
      this.createTile(letter, width, height);
    }

    return this.cache[letter];
  }

  // creates a canvas with a single letter
  // (for the fast font cache)
  createTile({text, tileWidth, tileHeight, tileFont, colorScheme}) {
    const key = text + "-" + colorScheme;
    if (key in this.cache) {
      return this.cache[key];
    }
    const canvas = this.cache[key] = document.createElement("canvas");
    canvas.width = tileWidth;
    canvas.height = tileHeight;
    this.ctx = canvas.getContext('2d');

    this.ctx.font = tileFont;
    this.ctx.globalAlpha = 0.7;
    this.ctx.fillStyle = colorScheme;
    this.ctx.fillRect(0, 0, tileWidth, tileHeight);
    this.ctx.globalAlpha = 1.0;

    this.ctx.fillStyle = "#000000";
    this.ctx.font = tileFont + "px mono";
    this.ctx.textBaseline = 'middle';
    this.ctx.textAlign = "center";
    this.ctx.fillText(text, tileWidth / 2, tileHeight / 2, tileWidth, tileFont);
    return canvas;
  }

  updateDimensions(width, height) {
    this.invalidate();
    this.cacheWidth = width;
    this.cacheHeight = height;
  }

  invalidate() {
    // TODO: destroy the old canvas elements
    this.cache = {};
  }
}

const cache = new CanvasCache();

/**
 * Allows rendering in tiles of grids.
 *
 * |---|---|---|
 * |-1-|-2-|-3-|
 * |---|---|---|
 * ―――――――――――――
 * |---|---|---|
 * |-4-|-5-|-6-|
 * |---|---|---|
 *
 * where 1..6 are TilingGrid component of xGridSize and yGridSize of 3.
 *
 * This split-up is required to avoid frequent repaints and keeps the React
 * Tree calculations slim.
 */
class CanvasTilingGridComponent extends CanvasComponent {

  drawTile({row, column}) {
    const tileWidth = this.props.tileWidth;
    const tileHeight = this.props.tileHeight;
    const yPos = tileHeight * (row - this.props.startYTile);
    const xPos = tileWidth * (column - this.props.startXTile);
    const el = this.props.sequences.raw[row].sequence[column];
    if (el !== undefined) {
      const colorScheme = this.props.colorScheme.getColor(el)
      const canvasTile = cache.createTile({
        colorScheme,
        text: el,
        tileWidth, tileHeight,
        tileFont: this.props.tileFont,
      });
      this.ctx.ctx.drawImage(
        canvasTile, xPos, yPos, tileWidth, tileHeight
      );
    }
  }

  draw() {
    const residues = [];
    for (let i = this.props.startYTile; i < this.props.endYTile; i++) {
      for (let j = this.props.startXTile; j < this.props.endXTile; j++) {
        this.drawTile({row:i, column:j});
      }
    }
  }

  render() {
    const style = {
      position: "absolute",
      top: 20 * this.props.startYTile,
      left: 20 * this.props.startXTile,
    }
    return (
      <div style={style}>
        <canvas
          ref={this.canvas}
          width={this.props.width}
          height={this.props.height}
        >
        </canvas>
      </div>
    );
  }
}

CanvasTilingGridComponent.defaultProps = {
  width: 200,
  height: 200,
}

CanvasTilingGridComponent.propTypes = {
  style: PropTypes.object,
  renderTile: PropTypes.func.isRequired,
  startXTile: PropTypes.number.isRequired,
  startYTile: PropTypes.number.isRequired,
  endXTile: PropTypes.number.isRequired,
  endYTile: PropTypes.number.isRequired,
};

export default CanvasTilingGridComponent;
