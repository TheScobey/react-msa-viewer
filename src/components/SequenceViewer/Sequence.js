/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { PureComponent } from 'react';

/**
 * Renders an individual sequence fragment.
 */
class Sequence extends PureComponent {
    render() {
      const {
        xPos,
        jPos,
        sequence,
        tileWidth,
        tileHeight,
        tileFont,
        colorScheme,
        width,
        residueComponent,
      } = this.props;
      const rawSequence = sequence.sequence;
      const residues = [];
      let xPosMoved = xPos;
      const Residue = residueComponent;
      for (let j = jPos; j < rawSequence.length; j++) {
        const el = rawSequence[j];
        residues.push(<Residue
          width={tileWidth}
          height={tileHeight}
          color={colorScheme.getColor(el)}
          font={tileFont}
          name={el}
          key={j}
        />);
        xPosMoved += tileWidth;
        if (xPosMoved > width)
            break;
      }
      return (
        <div>
          {residues}
        </div>
      );
    }
}
export default Sequence;
