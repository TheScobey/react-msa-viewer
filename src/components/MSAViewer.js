/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import MSAProvider from '../store/provider';

import PositionBar from './PositionBar';
import SequenceViewer from './SequenceViewer';
import SequenceOverview from './SequenceOverview';
import OverviewBar from './OverviewBar';
import Labels from './Labels';

import createMSAStore from '../store/createMSAStore';
import propsToRedux from '../store/propsToRedux';

const labelsAndSequenceDiv = {
  display: "flex",
};

// TODO: support changing the store dynamically
// TODO: when props of children update -> update store
// TODO: support using the child components in stand-alone mode
class MSAViewerComponent extends Component {

  componentWillMount() {
    //this.store = this.props.store || createMSAStore(
      //pick(this.props, MSAViewerProps)
    //);
  }
  render() {
    const {children, msaStore, ...otherProps} = this.props;
    if (children) {
      console.log("children: ", children);
      return (
        <MSAProvider store={msaStore}>
          <div {...otherProps}>
            {children}
          </div>
        </MSAProvider>
      );
    } else {
      // TODO: add more advanced layouts
      const currentState = msaStore.getState();
      const labelsPadding = currentState.viewpoint.tileSizes[1];
      const overviewBarHeight = currentState.viewpoint.overviewBar.height;
      const labelsStyle = {
        paddingTop: labelsPadding + overviewBarHeight,
      }
      const separatorPadding = {
        height: 10,
      };
      return (
        <MSAProvider store={msaStore}>
          <div style={labelsAndSequenceDiv}>
            <Labels
              style={labelsStyle}
            />
            <div>
              <OverviewBar />
              <PositionBar />
              <SequenceViewer />
              <div style={separatorPadding} />
              <SequenceOverview />
            </div>
          </div>
        </MSAProvider>
      );
    }
  }
}

const MSAViewer = propsToRedux(MSAViewerComponent);

MSAViewer.PropTypes = {
  // store
  // sequences, required
};

// TODO: re-include propsToRedux here?
export default MSAViewer;
export {
  createMSAStore,
  Labels,
  MSAViewer,
  OverviewBar,
  PositionBar,
  SequenceOverview,
  SequenceViewer,
};
