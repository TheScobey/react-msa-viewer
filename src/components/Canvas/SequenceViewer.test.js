/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React from 'react';
import ReactDOM from 'react-dom';

import SequenceViewer from './SequenceViewer';
import { SequenceViewer as CanvasSequenceViewer } from './SequenceViewer';
import MSAViewer from '../MSAViewer';
import {
  dummySequences
} from '../../test';

import { mount } from 'enzyme';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MSAViewer sequences={[...dummySequences]}>
    <SequenceViewer />
  </MSAViewer>, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('sends movement actions on mousemove events', () => {
  const msa = mount(<MSAViewer sequences={[...dummySequences]}
    width={400} height={200}
    >
    <SequenceViewer />
  </MSAViewer>);
  const el = msa.find('canvas').first();
  it('should have correctly rendered the canvas', () => {
    const props = el.props();
    expect(props.width).toBe(400);
    expect(props.height).toBe(140); // only 7 dummy sequences
  });

  it('should have correctly rendered the parent div', () => {
    const props = el.parent().props();
    expect(props.width).toBe(400);
    expect(props.height).toBe(140); // only 7 dummy sequences
  });
})

it("should fire an event on mouseclick", () => {
  const mockOnClick = jest.fn();
  const msa = mount(<MSAViewer
    sequences={[...dummySequences]}
    width={400} height={200}
    >
    <SequenceViewer onResidueClick={mockOnClick} />
  </MSAViewer>);
  expect(msa).toMatchSnapshot();
  const sv = msa.find(CanvasSequenceViewer).instance();
  const fakeClickEvent = {
    offsetX: 50,
    offsetY: 20,
  };
  sv.onClick(fakeClickEvent);
  expect(mockOnClick.mock.calls.length).toBe(1);
  expect(mockOnClick.mock.calls[0][0]).toEqual({
    "i": 1, "position": 2, "residue": "E",
    "sequence": {
      "name": "sequence 2",
      "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP"
    }
  });
})
