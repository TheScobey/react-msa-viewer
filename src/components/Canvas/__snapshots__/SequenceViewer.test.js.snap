// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fire an event on mouseclick 1`] = `
<PropsToReduxComponent
  backgroundColor="red"
  colorScheme="clustal"
  engine="canvas"
  height={200}
  sequences={
    Array [
      Object {
        "name": "sequence 1",
        "sequence": "MEEPQSDPSIEP-PLSQETFSDLWKLLPENNVLSPLPS-QA-VDDLMLSPDDLAQWLTED",
      },
      Object {
        "name": "sequence 2",
        "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
      },
      Object {
        "name": "sequence 3",
        "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
      },
      Object {
        "name": "sequence 4",
        "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
      },
      Object {
        "name": "sequence 5",
        "sequence": "MEEPQSD--IEL-PLSEETFSDLWWPLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
      },
      Object {
        "name": "sequence 6",
        "sequence": "MEEPQEDLSSSL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
      },
      Object {
        "name": "sequence 7",
        "sequence": "MEEPQ---SISE-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE---LSENVAGWLEDP",
      },
    ]
  }
  tileHeight={20}
  tileWidth={20}
  width={400}
>
  <MSAViewerComponent
    msaStore={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <div>
        <Connect(SequenceViewerComponent)
          onResidueClick={[MockFunction]}
        >
          <SequenceViewerComponent
            border={false}
            borderColor="black"
            borderWidth={1}
            cacheElements={20}
            colorScheme={
              ColorScheme {
                "scheme": StaticSchemeClass {
                  "defaultColor": "#ffffff",
                  "getColor": [Function],
                  "map": Object {
                    "A": "orange",
                    "B": "#fff",
                    "C": "green",
                    "D": "red",
                    "E": "red",
                    "F": "blue",
                    "G": "orange",
                    "Gap": "#fff",
                    "H": "red",
                    "I": "green",
                    "J": "#fff",
                    "K": "red",
                    "L": "green",
                    "M": "green",
                    "N": "#fff",
                    "O": "#fff",
                    "P": "orange",
                    "Q": "#fff",
                    "R": "red",
                    "S": "orange",
                    "T": "orange",
                    "U": "#fff",
                    "V": "green",
                    "W": "blue",
                    "X": "#fff",
                    "Y": "blue",
                    "Z": "#fff",
                  },
                  "type": "static",
                },
              }
            }
            dispatch={[Function]}
            fullHeight={140}
            fullWidth={1200}
            height={140}
            nrXTiles={21}
            nrYTiles={11}
            onResidueClick={[MockFunction]}
            overflow="hidden"
            overflowX="auto"
            overflowY="auto"
            scrollBarPositionX="bottom"
            scrollBarPositionY="right"
            sequences={
              Object {
                "length": 7,
                "maxLength": 60,
                "raw": Array [
                  Object {
                    "name": "sequence 1",
                    "sequence": "MEEPQSDPSIEP-PLSQETFSDLWKLLPENNVLSPLPS-QA-VDDLMLSPDDLAQWLTED",
                  },
                  Object {
                    "name": "sequence 2",
                    "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
                  },
                  Object {
                    "name": "sequence 3",
                    "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
                  },
                  Object {
                    "name": "sequence 4",
                    "sequence": "MEEPQSDLSIEL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
                  },
                  Object {
                    "name": "sequence 5",
                    "sequence": "MEEPQSD--IEL-PLSEETFSDLWWPLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
                  },
                  Object {
                    "name": "sequence 6",
                    "sequence": "MEEPQEDLSSSL-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE-LFLSENVAGWLEDP",
                  },
                  Object {
                    "name": "sequence 7",
                    "sequence": "MEEPQ---SISE-PLSQETFSDLWKLLPPNNVLSTLPS-SDSIEE---LSENVAGWLEDP",
                  },
                ],
              }
            }
            showModBar={false}
            textColor="black"
            textFont="18px Arial"
            tileHeight={20}
            tileWidth={20}
            width={400}
            xGridSize={10}
            yGridSize={10}
          >
            <div
              height={140}
              style={
                Object {
                  "cursor": "grab",
                  "height": 140,
                  "position": "relative",
                  "width": 400,
                }
              }
              width={400}
            >
              <canvas
                height={140}
                style={
                  Object {
                    "left": 0,
                    "position": "absolute",
                    "top": 0,
                  }
                }
                width={400}
              >
                Your browser does not seem to support HTML5 canvas.
              </canvas>
              <canvas
                height={140}
                style={
                  Object {
                    "left": 0,
                    "position": "absolute",
                    "top": 0,
                  }
                }
                width={400}
              >
                Your browser does not seem to support HTML5 canvas.
              </canvas>
              <FakeScroll
                fullHeight={140}
                fullWidth={1200}
                height={140}
                overflow="hidden"
                overflowX="auto"
                overflowY="auto"
                positionX="bottom"
                positionY="right"
                scrollBarWidth={5}
                width={400}
              >
                <div />
              </FakeScroll>
            </div>
          </SequenceViewerComponent>
        </Connect(SequenceViewerComponent)>
      </div>
    </Provider>
  </MSAViewerComponent>
</PropsToReduxComponent>
`;
